x/staking/simulation/operations.go

```diff
@@ -0,0 +1,401 @@
+			simulation.RandStringOfLength(r, 10),
+			simulation.RandStringOfLength(r, 10),
+			simulation.RandStringOfLength(r, 10),
+		)
+
+		msg := types.NewMsgEditValidator(address, description, &newCommissionRate, nil)
+
+		tx := helpers.GenTx(
+			[]sdk.Msg{msg},
+			fees,
+			chainID,
+			[]uint64{account.GetAccountNumber()},
+			[]uint64{account.GetSequence()},
+			[]crypto.PrivKey{simAccount.PrivKey}...,
+		)
+
+		res := app.Deliver(tx)
+		if !res.IsOK() {
+			return simulation.NoOpMsg(types.ModuleName), nil, errors.New(res.Log)
+		}
+
+		return simulation.NewOperationMsg(msg, true, ""), nil, nil
+	}
+}
+
+// SimulateMsgDelegate generates a MsgDelegate with random values
+// nolint: funlen
+func SimulateMsgDelegate(ak types.AccountKeeper, k keeper.Keeper) simulation.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simulation.Account, chainID string,
+	) (simulation.OperationMsg, []simulation.FutureOperation, error) {
+
+		denom := k.GetParams(ctx).BondDenom
+		if len(k.GetAllValidators(ctx)) == 0 {
+			return simulation.NoOpMsg(types.ModuleName), nil, nil
+		}
+
+		simAccount, _ := simulation.RandomAcc(r, accs)
+		val, ok := keeper.RandomValidator(r, k, ctx)
+		if !ok {
+			return simulation.NoOpMsg(types.ModuleName), nil, nil
+		}
+
+		if val.InvalidExRate() {
+			return simulation.NoOpMsg(types.ModuleName), nil, nil
+		}
+
+		amount := ak.GetAccount(ctx, simAccount.Address).GetCoins().AmountOf(denom)
+		if !amount.IsPositive() {
+			return simulation.NoOpMsg(types.ModuleName), nil, nil
+		}
+
+		amount, err := simulation.RandPositiveInt(r, amount)
+		if err != nil {
+			return simulation.NoOpMsg(types.ModuleName), nil, err
+		}
+
+		bondAmt := sdk.NewCoin(denom, amount)
+
+		account := ak.GetAccount(ctx, simAccount.Address)
+		coins := account.SpendableCoins(ctx.BlockTime())
+
+		var fees sdk.Coins
+		coins, hasNeg := coins.SafeSub(sdk.Coins{bondAmt})
+		if !hasNeg {
+			fees, err = simulation.RandomFees(r, ctx, coins)
+			if err != nil {
+				return simulation.NoOpMsg(types.ModuleName), nil, err
+			}
+		}
+
+		msg := types.NewMsgDelegate(simAccount.Address, val.GetOperator(), bondAmt)
+
+		tx := helpers.GenTx(
+			[]sdk.Msg{msg},
+			fees,
+			chainID,
+			[]uint64{account.GetAccountNumber()},
+			[]uint64{account.GetSequence()},
+			[]crypto.PrivKey{simAccount.PrivKey}...,
+		)
+
+		res := app.Deliver(tx)
+		if !res.IsOK() {
+			return simulation.NoOpMsg(types.ModuleName), nil, errors.New(res.Log)
+		}
+
+		return simulation.NewOperationMsg(msg, true, ""), nil, nil
+	}
+}
+
+// SimulateMsgUndelegate generates a MsgUndelegate with random values
+// nolint: funlen
+func SimulateMsgUndelegate(ak types.AccountKeeper, k keeper.Keeper) simulation.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accs []simulation.Account, chainID string,
+	) (simulation.OperationMsg, []simulation.FutureOperation, error) {
+
+		simAccount, idx := simulation.RandomAcc(r, accs)
```

Comment:
What is going on here??? Why not pick a validator, get all delegations and then pick a random delegation?